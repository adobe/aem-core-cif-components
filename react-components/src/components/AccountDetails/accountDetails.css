.root {
    display: grid;
    padding: 2.5rem 3rem;
    row-gap: 2rem;
    width: 100%;
    text-align: center;
    padding: 1rem;
    margin-top: 3rem;
}
.lineItems {
    display: grid;
    grid-gap: 0.75rem;
    grid-template-columns: 1fr auto;
    line-height: 1.5rem;
}

.lineItemLabel {
    font-weight: var(--venia-global-fontWeight-bold);
}

.editInformationButton {
    composes: root_normalPriority from '../Button/button.css';
    margin: auto;
}

.messageText {
    text-align: center;
    margin: auto;
}

@media (min-width: 961px) {
    .accountDetails {
        display: grid;
        gap: 1rem;
        grid-template-columns: 1fr 1fr 1fr;
        margin: 3rem 1rem;
    }
}

@media (max-width: 960px) {
    .lineItems {
        grid-template-columns: 1fr;
    }

    .lineItemButton {
        text-align: center;
    }
}
