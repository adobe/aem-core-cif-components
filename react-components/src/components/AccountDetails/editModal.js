/*******************************************************************************
 *
 *    Copyright 2020 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

import React from 'react';
import { object, bool, func, array } from 'prop-types';
import { useTranslation } from 'react-i18next';
import { useConfigContext } from '../../context/ConfigContext';
import Dialog from '../Dialog';
import FormError from '../FormError';

import EditForm from './editForm';

const EditModal = props => {
    const {
        initialValues,
        isOpen,
        onCancel,
        onSubmit,
        isDisabled,
        onChangePassword,
        shouldShowNewPassword,
        formErrors
    } = props;
    const [t] = useTranslation('account');

    const {
        mountingPoints: { accountDetails: rootSelector }
    } = useConfigContext();

    const dialogFormProps = { initialValues };

    return (
        <Dialog
            confirmText="Save"
            formProps={dialogFormProps}
            isOpen={isOpen}
            onCancel={onCancel}
            onConfirm={onSubmit}
            shouldDisableAllButtons={isDisabled}
            shouldDisableConfirmButton={isDisabled}
            title={t('account:edit-account-info', 'Edit account information')}
            rootContainerSelector={rootSelector}
            isModal={true}>
            <FormError errors={formErrors} />
            <EditForm handleShowNewPasswordField={onChangePassword} shouldShowNewPassword={shouldShowNewPassword} />
        </Dialog>
    );
};

export default EditModal;

EditModal.propTypes = {
    initialValues: object.isRequired,
    isOpen: bool,
    onCancel: func.isRequired,
    onSubmit: func.isRequired,
    isDisabled: bool,
    onChangePassword: func.isRequired,
    shouldShowNewPassword: bool,
    formErrors: array
};
