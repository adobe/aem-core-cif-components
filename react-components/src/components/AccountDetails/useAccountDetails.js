/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import { useMemo, useCallback, useState } from 'react';
import { useQuery, useMutation } from '@apollo/client';
import { useUserContext } from '../../context/UserContext';

const useAccountDetails = props => {
    const { getCustomerInformationQuery, setCustomerInformationMutation, changeCustomerPasswordMutation } = props;
    const [{ isSignedIn }] = useUserContext();

    // control the whether the edit form is shown or not
    const [isUpdateMode, setIsUpdateMode] = useState(false);

    // control whether we show the "new password" field
    const [shouldShowNewPasswordField, setShowNewPassword] = useState(false);

    // control whether we show the error or not.
    const [displayError, setDisplayError] = useState(false);

    const { data: accountInformationData } = useQuery(getCustomerInformationQuery, {
        skip: !isSignedIn,
        fetchPolicy: 'cache-and-network',
        nextFetchPolicy: 'cache-first'
    });

    const [
        setCustomerInformation,
        { loading: isUpdatingCustomer, error: customerInformationUpdateError }
    ] = useMutation(setCustomerInformationMutation, { refetchQueries: ['GetCustomerInformation'] });

    const [changeCustomerPassword, { loading: isChangingCustomerPassword, error: changePasswordError }] = useMutation(
        changeCustomerPasswordMutation
    );

    const initialValues = useMemo(() => {
        if (accountInformationData) {
            return {
                customer: accountInformationData.customer
            };
        }
    }, [accountInformationData]);

    const showEditForm = useCallback(() => {
        setIsUpdateMode(true);
        // hide the previous error message, if any
        setDisplayError(false);
    }, [setIsUpdateMode]);

    const handleCancel = useCallback(() => {
        setIsUpdateMode(false);
        setShowNewPassword(false);
    }, [setIsUpdateMode, setShowNewPassword]);

    const handleShowNewPasswordField = useCallback(() => {
        setShowNewPassword(true);
    }, [setShowNewPassword]);

    const handleSubmit = useCallback(
        async values => {
            const { firstname, lastname, email, password, newPassword } = values;
            try {
                if (
                    initialValues.customer.firstname !== firstname ||
                    initialValues.customer.lastname !== lastname ||
                    initialValues.customer.email !== email
                ) {
                    // we have to send the password when we change the email address
                    await setCustomerInformation({
                        variables: {
                            firstname,
                            lastname,
                            email,
                            password
                        }
                    });
                }
                // if we have a "new password" value we must change the password as well
                if (password && newPassword) {
                    await changeCustomerPassword({
                        variables: {
                            currentPassword: password,
                            newPassword
                        }
                    });
                }
                handleCancel();
            } catch {
                setDisplayError(true);
                return;
            }
        },
        [setCustomerInformation, handleCancel, changeCustomerPassword, initialValues]
    );

    const errors = displayError ? [customerInformationUpdateError, changePasswordError] : [];

    return {
        initialValues,
        formErrors: errors,
        isSignedIn,
        isUpdateMode,
        isDisabled: isUpdatingCustomer || isChangingCustomerPassword,
        showEditForm,
        handleCancel,
        handleSubmit,
        shouldShowNewPasswordField,
        handleShowNewPasswordField
    };
};

export default useAccountDetails;
