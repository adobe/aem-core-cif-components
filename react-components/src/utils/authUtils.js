/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import { ApolloLink } from '@apollo/client';
import { checkCookie, cookieValue } from './cookieUtils';

const authLink = new ApolloLink((operation, forward) => {
    let token = checkCookie('cif.userToken') ? cookieValue('cif.userToken') : '';
    if (token.length > 0) {
        operation.setContext(({ headers }) => ({
            headers: {
                authorization: `Bearer ${token && token.length > 0 ? token : ''}`,
                ...headers
            }
        }));
    }

    return forward(operation);
});

export { authLink as graphqlAuthLink };
