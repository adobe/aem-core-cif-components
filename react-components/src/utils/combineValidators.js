/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

/**
 * The function purpose is to group validation callbacks into a chain within single callback function.
 *
 * The single callback function should be passed as `validation` prop to an input component. It's being
 * called by the React Controller on a form submit.
 * The `callbacks` param is the array contains validators. Each validator should be either function or array. In case of a
 * validator is a function it is called the same way as React Controller does, with a field `value` and a form `values` params. If
 * the validator is an array, this means that it is extended validator which requires additional param for configuration.
 * The first index of that array should be a extended validator function, the second index is extended param, which is needed for
 * that function. So the function is called with three params on the form submit, with a field `value`, a form `values`
 * and an `extended param`.
 *
 * Example usage of the function:
 *
 * <input validate={combine([
 * foo
 * ])} .../>
 *
 * foo - validation function
 * It will be called on form submitting within chain as:
 * `foo(value, values)`
 *
 * Example usage of the function with extended callback:
 *
 * <input validate={combine([
 * [foo, bar]
 * ])} .../>
 *
 * foo - extended validator
 * bar - additional param
 * It will be called on form submitting within chain as:
 * `foo(value, values, bar)`
 *
 * Each callback being called one after another, according to their index in callbacks array, if one item returned the error
 * message, validation is failed, and rest validator are not supposed to be called.
 *
 * @param {Array} callbacks
 * @return {function(value, values): *}
 */
export default callbacks => {
    if (callbacks == null || !Array.isArray(callbacks)) {
        throw new Error('Expected `callbacks` to be array.');
    }

    return (value, values) => {
        let result = null;

        for (let i = 0; i < callbacks.length; i++) {
            const callback = callbacks[i];

            if (callback == null || (!Array.isArray(callback) && typeof callback !== 'function')) {
                throw new Error('Expected `callbacks[' + i + ']` to be array or function.');
            }

            if (Array.isArray(callback)) {
                const [extendedCallback, extendedParam] = callback;

                if (typeof extendedCallback !== 'function') {
                    throw new Error('Expected `callbacks[' + i + '][0]` to be function.');
                }

                result = extendedCallback(value, values, extendedParam);
            } else {
                result = callback(value, values);
            }

            if (result) {
                break;
            }
        }

        return result;
    };
};
