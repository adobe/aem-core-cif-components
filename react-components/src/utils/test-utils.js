/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { render } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { I18nextProvider } from 'react-i18next';

import i18n from '../../__mocks__/i18nForTests';
import ConfigContextProvider from '../context/ConfigContext';
import UserContextProvider from '../context/UserContext';

import mutationChangePassword from './mocks/mutationChangePassword';
import mutationCreateCustomer from './mocks/mutationCreateCustomer';
import mutationCreateDuplicateCustomer from './mocks/mutationCreateCustomerDuplicate';
import mutationDeleteCustomerAddress from './mocks/mutationDeleteCustomerAddress';
import mutationGenerateToken from './mocks/mutationGenerateToken';
import mutationGenerateTokenWrongPassword from './mocks/mutationGenerateTokenWrongPassword';
import mutationMergeCarts from './mocks/mutationMergeCarts';
import mutationPlaceOrder from './mocks/mutationPlaceOrder';
import mutationRevokeToken from './mocks/mutationRevokeToken';
import mutationShippingAddress from './mocks/mutationShippingAddress';
import queryCart from './mocks/queryCart';
import queryCountries from './mocks/queryCountries';
import queryCustomerCart from './mocks/queryCustomerCart';
import queryCustomerDetails from './mocks/queryCustomerDetails';
import queryCustomerInformation from './mocks/queryCustomerInformation';
import queryEmptyCart from './mocks/queryEmptyCart';
import queryNewCart from './mocks/queryNewCart';

const defaultMocks = [
    mutationChangePassword,
    mutationCreateCustomer,
    mutationCreateDuplicateCustomer,
    mutationDeleteCustomerAddress,
    mutationGenerateToken,
    mutationGenerateTokenWrongPassword,
    mutationMergeCarts,
    mutationPlaceOrder,
    mutationRevokeToken,
    mutationShippingAddress,
    queryCart,
    queryCountries,
    queryCustomerCart,
    queryCustomerDetails,
    queryCustomerInformation,
    queryEmptyCart,
    queryNewCart
];

const defaultConfig = {
    storeView: 'default',
    graphqlEndpoint: 'none'
};

// eslint-disable-next-line react/display-name
const allProviders = (config, userContext, mocks) => ({ children }) => {
    return (
        <MockedProvider mocks={mocks} addTypename={false}>
            <ConfigContextProvider config={config || defaultConfig}>
                <UserContextProvider initialState={userContext}>
                    <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
                </UserContextProvider>
            </ConfigContextProvider>
        </MockedProvider>
    );
};

/* Wrap all the React components tested with the library in a mocked Apollo provider */
const customRender = (ui, options = {}) => {
    const { config, userContext, mocks = defaultMocks, ...renderOptions } = options;
    return render(ui, { wrapper: allProviders(config, userContext, mocks), ...renderOptions });
};

export * from '@testing-library/react';
export { customRender as render };
