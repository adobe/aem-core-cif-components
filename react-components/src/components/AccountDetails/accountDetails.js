/*******************************************************************************
 *
 *    Copyright 2020 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React, { Fragment } from 'react';
import { useTranslation } from 'react-i18next';
import LoadingIndicator from '../LoadingIndicator';

import useAccountDetails from './useAccountDetails';
import GET_CUSTOMER_INFORMATION from './query_get_customer_information.graphql';
import classes from './accountDetails.css';

const AccountDetails = () => {
    const [t] = useTranslation('account');
    const { initialValues, isSignedIn } = useAccountDetails({ getCustomerInformation: GET_CUSTOMER_INFORMATION });

    if (!isSignedIn) {
        return (
            <div className={classes.messageText}>
                {t('account:address-book-sign-in-text', 'Please Sign in to see the account details.')}
            </div>
        );
    }

    if (!initialValues) {
        return <LoadingIndicator />;
    }
    const { customer } = initialValues;

    return (
        <div className={classes.accountDetails}>
            <div className={classes.lineItems}>
                <span className={classes.lineItemLabel}>{'Name'}</span>
                <span className={classes.lineItemValue}>{customer.firstname}</span>
                <span className={classes.lineItemLabel}>{'Email'}</span>
                <span className={classes.lineItemValue}>{customer.email}</span>
            </div>
        </div>
    );
};

export default AccountDetails;
