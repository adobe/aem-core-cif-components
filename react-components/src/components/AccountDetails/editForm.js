/*******************************************************************************
 *
 *    Copyright 2020 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

import React, { Fragment, useEffect } from 'react';
import { bool, func } from 'prop-types';
import { useTranslation } from 'react-i18next';
import { useFormApi, useFormState } from 'informed';

import Field from '../Field';
import LinkButton from '../LinkButton';
import TextInput from '../TextInput';
import Password from '../Password';
import combine from '../../utils/combineValidators';
import { isRequired, validatePassword, hasLengthAtLeast, isNotEqualToField } from '../../utils/formValidators';

import classes from './editForm.css';

const EditForm = props => {
    const { shouldShowNewPassword, handleShowNewPasswordField } = props;
    const [t] = useTranslation('account');

    const maybeShowChangePasswordButton = !shouldShowNewPassword && (
        <div className={classes.changePasswordButtonContainer}>
            <LinkButton
                type="button"
                classes={{ root: classes.changePasswordButton }}
                onClick={handleShowNewPasswordField}>
                {t('account:change-password', 'Change Password')}
            </LinkButton>
        </div>
    );

    const maybeShowNewPasswordField = shouldShowNewPassword ? (
        <div className={classes.newPassword}>
            <Password
                fieldName="newPassword"
                label={t('account:new-password', 'New Password')}
                validate={combine([
                    isRequired,
                    [hasLengthAtLeast, 8],
                    validatePassword,
                    [isNotEqualToField, 'password']
                ])}
                isToggleButtonHidden={false}
            />
        </div>
    ) : null;
    const passwordLabel = shouldShowNewPassword
        ? t('account:current-password', 'Current Password')
        : t('account:password', 'Password');

    const formState = useFormState();
    const formApi = useFormApi();
    const { submits, invalid } = formState;

    useEffect(() => {
        if (!invalid) {
            // clear the password field on a successful submit
            formApi.setValue('password', '');
        }
    }, [submits, invalid]);

    return (
        <Fragment>
            <div className={classes.root}>
                <div className={classes.firstname}>
                    <Field id="firstname" label={t('account:firstname', 'First Name')}>
                        <TextInput field="firstname" validate={isRequired} />
                    </Field>
                </div>
                <div className={classes.lastname}>
                    <Field id="lastname" label={t('account:lastname', 'Last Name')}>
                        <TextInput field="lastname" validate={isRequired} />
                    </Field>
                </div>
                <div className={classes.email}>
                    <Field id="email" label={t('account:email', 'Email address')}>
                        <TextInput field="email" validate={isRequired} />
                    </Field>
                </div>
                <div className={classes.password}>
                    <Password
                        fieldName="password"
                        label={passwordLabel}
                        validate={isRequired}
                        autoComplete="current-password"
                        isToggleButtonHidden={false}
                    />
                </div>
                {maybeShowChangePasswordButton}
            </div>
            {maybeShowNewPasswordField}
        </Fragment>
    );
};

export default EditForm;

EditForm.propTypes = {
    shouldShowNewPassword: bool,
    handleShowNewPasswordField: func
};
