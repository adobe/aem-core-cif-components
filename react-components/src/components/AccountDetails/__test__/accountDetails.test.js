/*******************************************************************************
 *
 *    Copyright 2020 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { screen, waitForElement } from '@testing-library/react';
import { render } from '../../../utils/test-utils';
import AccountDetails from '../accountDetails';
import getDetailsQuery from '../../../queries/query_get_customer_information.graphql';

const config = {
    storeView: 'default',
    graphqlEndpoint: 'none',
    mountingPoints: { accountDetails: 'mock' }
};

describe('<AccountDetails>', () => {
    it('renders the Account Details data for an unauthenticated user', () => {
        const { queryByText } = render(<AccountDetails />, { config: config, userContext: { isSignedIn: false } });

        expect(queryByText('Please Sign in to see the account details.')).not.toBeUndefined();
    });

    it('renders the Account Details data for an authenticated user', async () => {
        const mocks = [
            {
                request: {
                    query: getDetailsQuery
                },
                result: {
                    data: { customer: { id: 1, firstname: 'Jane', lastname: 'Doe', email: 'jdoe@gmail.com' } }
                }
            }
        ];
        const { queryByText } = render(<AccountDetails />, {
            config: config,
            userContext: { isSignedIn: true },
            mocks: mocks
        });

        await waitForElement(() => screen.getByLabelText('name'));

        expect(queryByText('Jane')).not.toBeUndefined();
        expect(queryByText('Doe')).not.toBeUndefined();
        expect(queryByText('jdoe@gmail.com')).not.toBeUndefined();
    });
});
