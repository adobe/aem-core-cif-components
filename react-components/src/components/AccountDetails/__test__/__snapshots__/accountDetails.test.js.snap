// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccountDetails> renders the Account Details form 1`] = `[Function]`;
